python.d/mongodb.chart.py

```diff
@@ -649,7 +650,13 @@ def _get_data(self):
 
     def _create_connection(self):
         conn_vars = {'host': self.host, 'port': self.port}
-        if hasattr(MongoClient, 'server_selection_timeout'):
+        if self.uri:
+            try:
+                connection = MongoClient(self.uri)
+                server_status = connection.admin.command('serverStatus')
+            except PyMongoError as error:
+                return None, None, str(error)
+        elif hasattr(MongoClient, 'server_selection_timeout'):
             conn_vars.update({'serverselectiontimeoutms': self.timeout})
```

Comment:
> serverSelectionTimeoutMS: (integer) Controls how long (in milliseconds) the driver will wait to find an available, appropriate server to carry out a database operation; while it is waiting, multiple server monitoring operations may be carried out, each controlled by connectTimeoutMS. Defaults to 30000 (30 seconds)

i think `serverselectiontimeoutms` should be setted for `uri` connections too.

Reply:
Can do, ill look at it tomorrow.

Reply:
@0xicl33n i see in `MongoClient` documentation

>         :Parameters:
>           - `host` (optional): hostname or IP address or Unix domain socket
>             path of a single mongod or mongos instance to connect to, or a
>             mongodb URI, or a list of hostnames / mongodb URIs. If `host` is
>             an IPv6 literal it must be enclosed in '[' and ']' characters
>             following the RFC2732 URL syntax (e.g. '[::1]' for localhost).
>             Multihomed and round robin DNS addresses are **not** supported.

So what if replace

https://github.com/firehol/netdata/blob/f47b067b49ee22cc01ac351c7a7c83c136806986/python.d/mongodb.chart.py#L377

with

```python
self.host = self.configuration.get('uri') or self.configuration.get('host', '127.0.0.1')
```

It seems that this will be enough

Reply:
Actually you can use `host` parameter instead of `uri`. @Ferroin what do you think?

Reply:
i mean `mongodb connect URI's` already supported by current version. All you need is to specify uri as a `host` in the configuration file.

Reply:
I'm not quite certain that that's the case.  It depends on how MongoClient handles being passed a port in addition to a URI.

At https://github.com/firehol/netdata/blob/master/python.d/mongodb.chart.py#L651, we're unconditionally including the port in the parameters we're putting together to initiate the connection, and that is gauaranteed to be a set value (either the default MongoDB port, or whatever the user supplied).  Given that, we can only keep things as-is iff MongoClient ignores the port parameter when `host` is a URI that includes a port, otherwise we end up not properly supporting the URI syntax.

Reply:
> we're unconditionally including the port in the parameters

yeah, this should be tested ofc

Reply:
Ok, i looked at the pymongo code again (`MongoClient` class `python-pymongo 3.6.1-1`)

```python
        for entity in host:
            if "://" in entity:
                res = uri_parser.parse_uri(entity, port, warn=True)
                seeds.update(res["nodelist"])
                username = res["username"] or username
                password = res["password"] or password
                dbase = res["database"] or dbase
                opts = res["options"]
            else:
                seeds.update(uri_parser.split_hosts(entity, port))
```

seems we are safe to include `port`

Reply:
OK, assuming that it passes testing, that appears to be fine then.  The documentation still needs updated (and I think the changes made by this PR actually do a great job of that), but there should be no code changes needed (though maybe adding a warning if `host` looks like a URI and the config also specifies a value for `port`?).

Reply:
> The documentation still needs updated

yes

Reply:
So then it should be looking more like..

```python
        conn_vars = {'host': self.host, 'port': self.port}
        if self.uri:
            try:
                connection = MongoClient(self.uri,self.timeout)
```
?